### PR TITLE
[#1354] Allow disowning of code by using empty author tags

### LINKED ISSUE #1354: empty author tags not honored
image

Guys, in the example above, the empty @@author tag at line 18 should mean the lines below it are not to be attributed to anyone until we encounter another non-empty @@author tag. But looks like some lines are being attributed to an author.
Is this a bug in the current logic?

File https://github.com/AY2021S1-CS2103T-W12-1/tp/blob/master/src/main/java/seedu/address/storage/JsonPatientRecordsStorage.java

### PR DESCRIPTION
**Pull request by dcshzj**
Fixes #1354.
:warning: Requires https://github.com/reposense/testrepo-Alpha/pull/6 to be merged first and the test repository URL to be modified :warning:
Adapted from #1357.

```
Empty author tags are assumed to represent the end of an
annotation block. However, there may be cases where certain code
blocks should not be attributed to any author. The current
workaround is to attribute the code to some random author, which
is not ideal from the user's perspective.

Let's implement a new functionality for users to disown code by
using empty author tags.
```

Verification that it works:
- [RepoSense report based on old code](https://dcshzj.github.io/charlie-old/?search=&sort=groupTitle&sortWithin=title&timeframe=commit&mergegroup=&groupSelect=groupByRepos&breakdown=false&since=&tabOpen=true&tabType=authorship&tabAuthor=Wang%20Yiming&tabRepo=reposense%2Ftestrepo-Charlie%5Bmaster%5D&authorshipIsMergeGroup=false&authorshipFileTypes=java~md~adoc) - Refer to AddressBookParserTest.java file lines 121 - 123
- [RepoSense report based on this patch](https://dcshzj.github.io/charlie-new/?search=&sort=groupTitle&sortWithin=title&since=&timeframe=commit&mergegroup=&groupSelect=groupByRepos&breakdown=false&tabOpen=true&tabType=authorship&tabAuthor=Wang%20Yiming&tabRepo=reposense%2Ftestrepo-Charlie%5Bmaster%5D&authorshipIsMergeGroup=false&authorshipFileTypes=java~md~adoc&authorshipIsBinaryFileTypeChecked=false) - Same file but notice that lines 121 - 123 are missing
- [The original blame view for reference](https://github.com/reposense/testrepo-Charlie/blame/867d66eab9bd0ddc3656cb960b6a6fcabcd5e22e/src/test/java/seedu/address/logic/parser/AddressBookParserTest.java)

### CODE REVIEW

**Comment by HCY123902** [comment]
Can you generate a report for testrepo-Charlie to verify if it works? Thanks

**Comment by dcshzj** [comment]
> Actually, can you generate a report for testrepo-Charlie to verify if it works?

@HCY123902 I added the links in the PR description.

**Review by HCY123902 — APPROVED** [review]
LGTM

**Comment by gerhean** [comment]
So what will represent the end of an annotation block now?

**Comment by damithc** [comment]
> So what will represent the end of an annotation block now?

The requirement is, the code block continues until the next `@@author` tag or the end of file.

**Comment by 0blivious** [comment]
Is it possible to have test cases for this new feature?

**Comment by dcshzj** [comment]
> Is it possible to have test cases for this new feature?

I have added some tests in https://github.com/reposense/testrepo-Alpha/pull/6. However, it is now running on [my fork](https://github.com/dcshzj/testrepo-Alpha), so the following line needs to be modified before merging:

https://github.com/reposense/RepoSense/blob/c8e9c7b2a26f8f257ea735426eace88265a3d1f6/src/test/java/reposense/template/GitTestTemplate.java#L32

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java [hunk 1]
```diff
@@ -135,18 +135,18 @@ public void analyzeCommits_emailWithAdditionOperator_success() {
 
     @Test
     public void analyzeCommits_duplicateAuthorsDuplicateCommits_success() throws Exception {
```
Comment (Tejas2805) [inline src/test/java/reposense/commits/CommitInfoAnalyzerTest.java:137]: Why is the testing for empty author tag being done inside `duplicate`? Either create a new test or modify this test name.

**Review by Tejas2805 — COMMENTED** [review]
One nit.

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java (hunk 1 above)
Comment (dcshzj) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java:137]: Hmm, this test case isn't testing for the empty author tag. The code in this test was updated to match the latest commit information, because a new commit was added to add more lines to the annotationTest.java file. Thus, the `LATEST_COMMIT_HASH` has to be updated as well as this test.

Comment (Tejas2805) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java:137]: Oh so got updated with the fake author name, got it. Can add a test for testing the empty author tag.

Comment (dcshzj) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java:137]: This test class is mainly testing the analysis of commits, perhaps you were referring to `AnnotatorAnalyzerTest`? The two test cases there (`analyzeAnnotation_authorNamePresentInConfig_overrideAuthorship` and `analyzeAnnotation_authorNameNotInConfig_disownCode`) covers the empty author tag, which is based on the updated `annotationTest.java` file where I added some empty author tags.

Comment (dcshzj) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java:137]: @Tejas2805 Ping

Comment (Tejas2805) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java:137]: Sorry for the delayed response. Thanks for the ping.

**Review by Tejas2805 — COMMENTED** [review]
LGTM!

Just a quick question. The code that doesn't get attributed to any author? Can it still be seen in the dashboard?

**Comment by dcshzj** [comment]
> Just a quick question. The code that doesn't get attributed to any author? Can it still be seen in the dashboard as attributed to no author?

No, code that gets disowned will not be shown in the dashboard and will just be ignored. As we only display code by author, we do not have an option to display code owned by nobody.

**Comment by Tejas2805** [comment]
Sorry for the delay. LGTM!

**Review by Tejas2805 — APPROVED** [review]

**Comment by fzdy1914** [comment]
@dcshzj You may proceed to modify the url

**Comment by fzdy1914** [comment]
@dcshzj Ping, you can modify the url now.

**Comment by dcshzj** [comment]
@fzdy1914 Updated, apologies for the delay.

**Comment by fzdy1914** [comment]
This is a very good sample of how a Good PR should be.

**Comment by damithc** [comment]
> This is a very good sample of how a Good PR should be.

Good work @dcshzj 👍

File: src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java [hunk 2]
```diff
@@ -53,15 +52,23 @@ public static void aggregateAnnotationAuthorInfo(FileInfo fileInfo, Map<String,
...
+                    return Optional.of(Author.UNKNOWN_AUTHOR);
+                }
                 return Optional.empty();
             }
             return Optional.of(authorAliasMap.getOrDefault(name, Author.UNKNOWN_AUTHOR));
         } catch (ArrayIndexOutOfBoundsException e) {
+            if (!currentAnnotatedAuthor.isPresent()) {
+                return Optional.of(Author.UNKNOWN_AUTHOR);
+            }
             return Optional.empty();
```
Comment (damithc) [inline src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java:72]: for future reference, this may be a good place to use the ternary operator.

```java
return currentAnnotatedAuthor.isPresent() ? Optional.empty() : Optional.of(Author.UNKNOWN_AUTHOR);
```

Comment (fzdy1914) [reply src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java:72]: Yes, truly. However, many languages other than Java do not support ternary operators.
